File: src/net/dinkla/raytracer/materials/Emissive.java
package net.dinkla.raytracer.materials;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.colors.RGBColor;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.worlds.World;

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 14.04.2010
 * Time: 21:10:53
 * To change this template use File | Settings | File Templates.
 */
public class Emissive<C extends Color> extends Material<C> {

    public float ls;
    public C ce;

    protected C cachedLe;

    public Emissive() {
        ls = 1.0f;
        ce = (C) C.getWhite();
    }
   
    public Emissive(C ce, float ls) {
        this.ce = ce;
        this.ls = ls;
    }

    @Override
    public C shade(World world, Shade sr) {
        throw new RuntimeException("Emissive.shade");
    }
    
    @Override
    public C areaLightShade(World world, Shade sr) {
        if (sr.getNormal().negate().dot(sr.ray.d) > 0) {
            return getLe(sr);
        } else {
            return (C) C.getBlack();
        }
    }

    @Override
    public C getLe(Shade sr) {
        if (null == cachedLe) {
            cachedLe = (C) ce.mult(ls);
        }
        return cachedLe;
    }
    
}

File: src/main/kotlin/net/dinkla/raytracer/materials/Emissive.kt
package net.dinkla.raytracer.materials

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.utilities.equals
import net.dinkla.raytracer.world.World
import java.util.Objects

class Emissive(private val ce: Color = Color.WHITE, val ls: Double = 1.0) : IMaterial {

    private val le: Color = ce * ls

    override fun shade(world: World, sr: Shade): Color = le

    override fun areaLightShade(world: World, sr: Shade): Color = if ((-(sr.normal)) dot (sr.ray.direction) > 0) {
        le
    } else {
        Color.BLACK
    }

    override fun getLe(sr: Shade): Color = le

    override fun equals(other: Any?): Boolean = this.equals<Emissive>(other) { a, b ->
        a.ce == b.ce && a.ls == b.ls
    }

    override fun hashCode(): Int = Objects.hash(ce, ls)

    override fun toString() = "Emissive($ce, $ls)"
}
